fendermint/ipc/src/parent.rs

```diff
@@ -0,0 +1,202 @@
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> Option<Vec<u8>> {
+        self.read_cache(|cache| cache.block_hash.get_value(height).map(|v| v.to_vec()))
+    }
+
+    fn top_down_msgs(&self, _height: BlockHeight, nonce: Nonce) -> Vec<CrossMsg> {
+        self.read_cache(|cache| {
+            let v = cache.top_down_message.values_within_range(nonce, None);
+            // FIXME: avoid clone here, return references, let caller clone on demand
+            v.into_iter().cloned().collect()
+        })
+    }
+
+    fn membership(&self) -> Option<ValidatorSet> {
+        self.read_cache(|cache| cache.membership.clone())
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) {
+        let mut cache = self.cache.write().unwrap();
+        cache.block_hash.remove_key_till(finality.height);
+    }
+}
+
+impl Clone for PollingParentSyncer {
+    fn clone(&self) -> Self {
+        Self {
+            config: self.config.clone(),
+            started: Arc::new(AtomicBool::new(false)),
+            cache: self.cache.clone(),
+            ipc_agent_proxy: self.ipc_agent_proxy.clone(),
+            handle: None,
+        }
+    }
+}
+
+struct ParentSyncerCache {
+    block_hash: RangeKeyCache<BlockHeight, Vec<u8>>,
+    top_down_message: RangeKeyCache<Nonce, CrossMsg>,
+    membership: Option<ValidatorSet>,
+}
+
+type LockedCache = Arc<RwLock<ParentSyncerCache>>;
+
+impl PollingParentSyncer {
+    fn read_cache<F, T>(&self, f: F) -> T
+    where
+        F: Fn(&ParentSyncerCache) -> T,
+    {
+        let cache = self.cache.read().unwrap();
+        f(&cache)
+    }
+
+    /// Start the proof of finality listener in the background
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let parent_syncer = self.ipc_agent_proxy.clone();
+        let config = self.config.clone();
+        let cache = self.cache.clone();
+
+        let handle =
+            tokio::spawn(async move { sync_with_parent(config, parent_syncer, cache).await });
+        self.handle = Some(handle);
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent(
+    config: Config,
+    agent_proxy: Arc<AgentProxy>,
+    lock: LockedCache,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        let hashes = fetch_block_hashes(&config, &agent_proxy, &lock).await?;
+        update_top_down_msgs(&config, &agent_proxy, &lock).await?;
+        update_membership(&config, &agent_proxy, &lock).await?;
+
+        let mut cache = lock.write().unwrap();
+        for r in hashes {
+            cache.block_hash.insert_after_lower_bound(r.0, r.1);
+        }
```

Comment:
Is there a reason why hashes are inserted after all the other updates? Why not have `update_block_hashes` instead of `fetch_block_hashes`?